# 6.01 Finding Important Variables for Predicting Responses to a Marketing Offer:

You have been given data regarding a company's marketing campaign, wherein discounts were offered for various products. You are interested in building a model that predicts the number of responses to an offer. It should also provide information about how much of a discount the offer included, how many customers the offer reached, and the value representing the offer quality that the marketing team assigned to that offer. 

You want to build a model that is accurate but does not contain unnecessary variables. 

Use RMSE to evalulate how the model performs when all variables are included and compare this to what happens when each variable is dropped from the model. 

In [16]:
import pandas as pd

offer = pd.read_csv('Activity6.01/offer_responses.csv')

offer.head()

responses  offer_discount  offer_quality  offer_reach
0     4151.0            26.0      10.257680      31344.0
1     3397.0            35.0      15.194380      24016.0
2     3274.0            21.0      13.971468      28832.0
3     3426.0            27.0       6.054338      26747.0
4     5745.0            42.0      16.801365      46968.0

Import train_test_split from Sklearn and use it to split the data into training and test sets.

Use responses as y variable and all others as x predictor variable.

In [17]:
from sklearn.model_selection import train_test_split

X = offer[['offer_discount', 
          'offer_quality',
          'offer_reach']]

y = offer['responses']


X_train, X_test, y_train,  y_test = train_test_split(X, y, random_state=10)

Import Linear Regression and Mean_squared_error from sklearn.
Fit the model to the training data and print out the RMSE.


In [18]:
from sklearn.linear_model import LinearRegression

Lin_model = LinearRegression()

Lin_model.fit(X_train, y_train)

Lin_prediction = Lin_model.predict(X_test)


from sklearn.metrics import mean_squared_error

print('RMSE: ' + str(mean_squared_error(Lin_prediction, y_test)**.5) )

RMSE: 966.2461828577913


Drop offering_quality from the training data and run RMSE. 

In [19]:
X1 = offer[['offer_discount',
           'offer_reach']]

X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y, random_state =10)

In [20]:
Lin_model1 = LinearRegression()

Lin_model1.fit(X_train1, y_train1)

Lin_prediction1 = Lin_model1.predict(X_test1)

print('RMSE: '+str(mean_squared_error(Lin_prediction1, y_test1)**.5))

RMSE: 965.5346123758474


Drop offer_discount instead of offer_quality and run RMSE.

In [21]:
X2 = offer[['offer_quality',
           'offer_reach']]

X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y, random_state =10)

In [22]:
Lin_model2 = LinearRegression()

Lin_model2.fit(X_train2, y_train2)

Lin_prediction2 = Lin_model2.predict(X_test2)

print('RMSE: '+str(mean_squared_error(Lin_prediction2, y_test2)**.5))

RMSE: 1231.6766556327284


Drop offer_reach and run RMSE

In [23]:
X3 = offer[['offer_quality',
           'offer_discount']]

X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y, random_state =10)

In [24]:
Lin_model3 = LinearRegression()

Lin_model3.fit(X_train3, y_train3)

Lin_prediction3 = Lin_model3.predict(X_test3)

print('RMSE: '+str(mean_squared_error(Lin_prediction3, y_test3)**.5))

RMSE: 1185.8456831644116


RMSE Summary: 

- All variables: RMSE: 966.2461828577913
- Without offer_quality: RMSE: 965.5346123758474
- Without offer_discount: RMSE: 1231.6766556327284
- Without offer_reach: RMSE: 1185.8456831644116

The Change in RMSE value is directly related to the to the importance of the feature being removed. 
If RMSE goes down, the predictions of the model were more accurate. and feature can be removed. 
If RMSE goes up, the predictions were less accurate and feature belongs in the model. 

Without offer_quality the RMSE went down so this feature could be removed and simplify the model. 

### End Challenge Activity 6.01

# Challenge Activity 6.02: Using RFE To Choose Features For Predicting Customer Spend

You are given the following information (features) of various customers:
- prev_year_spend
- days_since_last_purchase
- days_since_first_purchase
- total_transactions
- age
- income
- engagement_score

You are tasked to investigate the relationship is to customer spend in the current year. Create a simple linear model to describe the relationships.

In [68]:
Cspend = pd.read_csv('Activity6.02/customer_spend.csv')

Cspend.head()

cur_year_spend  prev_year_spend  days_since_last_purchase  \
0         5536.46          1681.26                         7   
1          871.41          1366.74                        12   
2         2046.74          1419.38                        10   
3         4662.70          1561.21                        12   
4         3539.46          1397.60                        17   

   days_since_first_purchase  total_transactions  age    income  \
0                         61                  34   61  97914.93   
1                         34                  33   68  30904.69   
2                         81                  22   54  48194.59   
3                         32                  34   49  93551.98   
4                         72                  34   66  66267.57   

   engagement_score  
0         -0.652392  
1          0.007327  
2          0.221666  
3          1.149641  
4          0.835834

Use train_test_split with random_sate=100, cur_year_spend as y variable. 

In [69]:
Xcs = Cspend[['prev_year_spend',
             'days_since_last_purchase',
             'days_since_first_purchase',
             'total_transactions',
             'age',
             'income',
             'engagement_score']]

ycs =Cspend['cur_year_spend']

X_traincs , X_testcs, y_traincs, y_testcs = train_test_split(Xcs, ycs, random_state=100)

In [70]:
modelc = LinearRegression()
modelc.fit(X_traincs, y_traincs)

modelc_predict = modelc.predict(X_testcs)

print('RMSE: '+str(mean_squared_error(modelc_predict, y_testcs)**.5))

RMSE: 50.45039920074803


In [71]:
modelc.coef_

array([ 0.85320131, -0.3119968 , 15.55680368, 49.05766265, -0.0852239 ,
        0.05969942,  1.06228814])

Model is using all variables from non-zero values. We must simplify the model. 

Use RFE to obtain the 3 most important features and create new training and test datasets using only the selected columns.


In [72]:
from sklearn.feature_selection import RFE

rfe = RFE(estimator=LinearRegression(),n_features_to_select=3)

rfe.fit(X_traincs, y_traincs)

for featureselect in range(X_traincs.shape[1]):
    if rfe.support_[featureselect] == True:
    
        print('Feature: {}, Rank: {}'.format(X_traincs.columns[featureselect],
                                        rfe.ranking_[featureselect]))

Feature: days_since_first_purchase, Rank: 1
Feature: total_transactions, Rank: 1
Feature: engagement_score, Rank: 1


Rank score of 1 means that the rfe believes these features are equally important.

In [73]:
Xcs_rfe = Cspend[['days_since_first_purchase',
                 'total_transactions',
                 'engagement_score']]

X_trainrfe, X_testrfe, y_trainrfe, y_testrfe = train_test_split(Xcs_rfe, ycs, random_state=100)

Train a linear regression model with the simplified data. Calculate RSME. 


In [74]:
modelcs = LinearRegression()
modelcs.fit(X_trainrfe, y_trainrfe)

LinearRegression()

In [75]:
modelcs_predict = modelcs.predict(X_testrfe)

print('RMSE: '+str(mean_squared_error(modelcs_predict, y_testrfe)**.5))

RMSE: 1075.9083016269915


### End Challenge Activity 6.02


# Challenge Activity 6.03: Building the Best Regression Model for Customer Spend Based on Demographic Data

You are given customer spend and demographic data. Build a predictive model that can predict customer spend based on the demographic data. 

In [91]:
BRM = pd.read_csv('Activity6.03/spend_age_income_ed.csv')

BRM.head()

spend   age   income  years_of_education
0  3304.0  36.0  45125.0                  12
1  3709.0  43.0  41695.0                  10
2  3305.0  47.0  39253.0                  17
3  2170.0  33.0  32384.0                  13
4  2113.0  30.0  33182.0                  10

Perform a train_test_split with random_state=10.


In [93]:

Xbrm = BRM[['age',
           'income',
           'years_of_education']]

ybrm = BRM['spend']

X_trainbrm, X_testbrm, y_trainbrm, y_testbrm = train_test_split(Xbrm, ybrm, random_state=10)



Fit a linear Regression model to the training data. Print RMSE.

In [95]:
model_brm = LinearRegression()

model_brm.fit(X_trainbrm, y_trainbrm)

model_brm_prediction = model_brm.predict(X_testbrm)

print('RMSE: '+str(mean_squared_error(model_brm_prediction, y_testbrm)**.5))

RMSE: 348.19771532747865


Fit regression tree models to the data, with the max_depth = 2 & 5. Calculate RMSE

In [97]:
from sklearn.tree import DecisionTreeRegressor

tree_model_2 = DecisionTreeRegressor(max_depth = 2)
tree_model_2.fit(X_trainbrm, y_trainbrm)

tree_model_2_prediction = tree_model_2.predict(X_testbrm)
print('RMSE: '
    + str(mean_squared_error(tree_model_2_prediction, y_testbrm)**.5))

tree_model_5 = DecisionTreeRegressor(max_depth=5)
tree_model_5.fit(X_trainbrm, y_trainbrm)

tree_model_5_prediction = tree_model_5.predict(X_testbrm)
print('RMSE: '
    + str(mean_squared_error(tree_model_5_prediction, y_testbrm)**.5))


RMSE: 268.51069264082935
RMSE: 125.53257106419696


Fit Random Forest models to the data, with the max_depth = 2 & 5. Calculate RMSE

In [100]:
from sklearn.ensemble import RandomForestRegressor

Forest_model2 = RandomForestRegressor(max_depth = 2, random_state=10)
Forest_model2.fit(X_trainbrm, y_trainbrm)

Forest_model2_prediction = Forest_model2.predict(X_testbrm)
print('RMSE:'+
     str(mean_squared_error(Forest_model2_prediction,y_testbrm)**.5))



Forest_model5 = RandomForestRegressor(max_depth = 5, random_state=10)
Forest_model5.fit(X_trainbrm, y_trainbrm)

Forest_model5_prediction = Forest_model5.predict(X_testbrm)
print('RMSE:'+
     str(mean_squared_error(Forest_model5_prediction,y_testbrm)**.5))

RMSE:266.45844988320863
RMSE:115.2014058797442


Summary of RMSE by Regression Model:

- Linear Regression: RMSE: 348.19771532747865
- Decision Tree MD 2: RMSE: 268.51069264082935
- Decision Tree MD 5: RMSE: 125.53257106419696
- Random Forest MD 2: RMSE: 268.51069264082935
- Random Forest MD 5: RMSE: 125.53257106419696

The Random Forest with the max_depth of 5 performs the best out of all the models. 

### End of Activity 6.03 